spring-cloud-deployer-local/src/main/java/org/springframework/cloud/deployer/spi/local/LocalAppDeployer.java

```diff
@@ -175,6 +175,17 @@ public String deploy(AppDeploymentRequest request) {
 				appInstanceEnv.put("SPRING_APPLICATION_INDEX", Integer.toString(i));
 				appInstanceEnv.put("SPRING_CLOUD_APPLICATION_GUID", Integer.toString(port));
 
+				// we only set 'normal' style props reflecting what we set for env format
+				// for cross reference to work inside SAJ.
+				// looks like for now we can't remove these env style formats as i.e.
+				// DeployerIntegrationTestProperties in tests really assume 'INSTANCE_INDEX' and
+				// this might be indication that we can't yet fully remove those.
+				if (useSpringApplicationJson(request)) {
+					appInstanceEnv.put("instance.index", Integer.toString(i));
+					appInstanceEnv.put("spring.application.index", Integer.toString(i));
+					appInstanceEnv.put("spring.cloud.application.guid", Integer.toString(port));
+				}
+
```

Comment:
Shouldn't the other way of setting properties be in an else here?

Reply:
I put a comment to PR to read reason for that from comment in code as I was 100% sure someone would ask it. In nutshell, tests break if we take it away.

Reply:
So we should fix the test to be valid under the new conditions.  And since you bring it up, we should have a test for this scenario as well.

Reply:
Maybe that `spi-test-app` could be enhanced to track these. We're actually missing integration tests for metric collector as integration with that was only thing what got broken what this change is fixing.

Reply:
Why can't we validate these in a similar way to the way we do with the command builders like here: https://github.com/spring-cloud/spring-cloud-deployer-local/blob/master/spring-cloud-deployer-local/src/test/java/org/springframework/cloud/deployer/spi/local/JavaExecutionCommandBuilderTests.java#L168

Reply:
That sounds reasonable. It doesn't test the actual problem for spring resolving cross refs, afaik, but is better than nothing.

Reply:
I agree, should be an `else` cause as it just looks confusing if anyone starts to debug to see the same property set twice (once in env-var and once in SAJ).